File: problems/0201-0300/0201_bitwise-and-of-numbers-range/solution_common_prefix.java
class Solution {

    public int rangeBitwiseAnd(int left, int right) {
        int shift = 0;
        // Shift both endpoints right until they agree: what remains is the
        // common binary prefix. Every bit below it flips through 0 somewhere
        // in [left, right], so the range's AND keeps only the prefix.
        while (left < right) {
            left >>>= 1;
            right >>>= 1;
            shift++;
        }
        // Restore the prefix to its original position, zeros below.
        return left << shift;
    }
}

File: problems/0201-0300/0201_common-prefix-bits/solution_common_prefix.ts
function commonPrefixBits(left: number, right: number): number {
    let shift = 0;
    // Shift both endpoints right until they agree: what remains is the
    // common binary prefix. Every bit below it flips through 0 somewhere in
    // [left, right], so the range's AND keeps only the prefix.
    while (left < right) {
        left = Math.floor(left / 2);
        right = Math.floor(right / 2);
        shift++;
    }
    // Restore the prefix to its original position, zeros below.
    return left * Math.pow(2, shift);
}
